### PR TITLE
Device websocket changes

### PR DESCRIPTION
**Pull request by arctan5x**
- Device -> Audio Server ping uses a socket over a 5 second period to measure RTT.
- Device <-> API Server uses tcp socket when a device is connected to an audio server. Otherwise, back to old ping/config exchange. This allows only "connected" devices establish a socket so we don't overload NATs.

### CODE REVIEW

File: cmd/services.go [hunk 1]
```diff
@@ -257,7 +257,6 @@ func restartAllServices(config client.AgentConfig, isServer bool) {
 		err = startService(conn, serviceName)
 		if err != nil {
 			log.Error(err, "Unable to start service", "name", serviceName)
-			panic(err)
```
Comment (mikedickey) [inline cmd/services.go:260]: i'm not sure it's a good idea to ignore service start failures. currently, this causes the agent to exit and get restarted by systemd, which allows it to keep retrying to recover from any failures.

File: cmd/deviceWSConnector.go [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+}
+
+// Connector is used as a singleton object of webSocketConnector
+var Connector webSocketConnector
+
+// ConfigChannel holds config objects that came through a websocket connection and is consumed by a handler in device.go
+var ConfigChannel = make(chan client.AgentConfig, 100)
+
+// ReceivePingChannel is used to control ReceivePingHandler Goroutine
+var ReceivePingChannel chan bool
```
Comment (mikedickey) [inline cmd/deviceWSConnector.go:43]: similar to devicePinger, i think we can (and should) eliminate the singletons by moving the variables & methods into the `webSocketConnector` struct.

File: cmd/deviceWSConnector.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+// SendDevicePing sends a current ping statistics through Connector
+func SendDevicePing(pingStats client.PingStats) {
+	// encode ping content
+	pingBytes, err := json.Marshal(pingStats)
+	if err != nil {
+		log.Error(err, "Failed to marshal agent ping request")
+	}
+
+	// effectively let this Goroutine time out in 3 seconds
+	Connector.Conn.SetWriteDeadline(time.Now().Add(3 * time.Second))
```
Comment (mikedickey) [inline cmd/deviceWSConnector.go:121]: I think it's risky that you can have multiple threads/routines using this socket at the same time. I recommend having a single goroutine running inside this class that handles all send and receive operations, and uses channels to communicate with other threads.

**Review by mikedickey — CHANGES_REQUESTED** [review]

File: cmd/services.go [hunk 4]
```diff
@@ -286,10 +285,11 @@ func stopService(conn *dbus.Conn, u dbus.UnitStatus) error {
 
 // startService is used to start a managed systemd service
 func startService(conn *dbus.Conn, name string) error {
-	log.Info("Starting managed service", "service", name)
+	log.Info(fmt.Sprintf("Starting managed service %s", name))
```
[the remark this answers is not in the document]
Comment (arctan5x) [reply cmd/services.go:288]: it always has to be a pair.

**Comment by arctan5x** [comment]
@mikedickey I tried with timing out the read channel and use a case to write if there is any pending message. Gorilla websocket's setReadTimeout actually marks the connection corrupt when it times out, so we cannot use it in the loop since our websocket will be unusable every iteration in the for loop. Instead, I made two handlers with your favorite read and write locks ;)

**Review by nwang92 — APPROVED** [review]

**Review by nwang92 — APPROVED** [review]

**Review by mikedickey — APPROVED** [review]

**Pull request merged**